### ISSUE TITLE
Allow for route-registrar health check scripts to access job-specific config (such as a CA cert)

### ISSUE THREAD

**Issue by tcdowney** [cloudfoundry/routing-release#147]
## Issue

This issue pertains to the [route-registrar](https://github.com/cloudfoundry/route-registrar) and its configuration in routing-release.

## Context

On the CAPI team we're working on a story ([#164937832](https://www.pivotaltracker.com/story/show/164937832) to allow operators to disable all non-TLS endpoints for `capi-release` jobs.

The Cloud Controller API uses the route-registrar to register its public route (i.e. `api.<system-domain>`) with the Gorouters and it [configures the route-registrar](https://github.com/cloudfoundry/cf-deployment/blob/master/cf-deployment.yml#L969-L983) to use the following health check script:

* [/var/vcap/jobs/cloud_controller_ng/bin/cloud_controller_ng_health_check](https://github.com/cloudfoundry/capi-release/blob/ca291cef55753a07214adb6fa3a49dc2ef914425/jobs/cloud_controller_ng/templates/cloud_controller_api_health_check.erb)

Unfortunately that script is still using the non-TLS `9022` port, so when we disable our non-TLS endpoints it breaks route registration. Configuring it to hit our TLS-protected endpoint on `9024` is fine in general, but if the Cloud Controller is using a self-signed certificate we need some way to expose the CA to the health check script so that we can trust it.

That CA certificate is available on disk for the Cloud Controller job, but since route-registrar is running within BPM it is not able to access it (see the [bpm.yml](https://github.com/cloudfoundry/routing-release/blob/c97c316b1b08550e9eca1cb0c45df92528b42771/jobs/route_registrar/templates/bpm.yml.erb#L27) config).

**tl;dr**
What we're asking for is some way to provide a CA cert to the health check script so that it can successfully health check Cloud Controller over TLS. More generally, this might look like providing a way to either include additional config for the health check script or providing more configuration around the additional volumes BPM will mount for it.

## Expected result
The route-registrar enables us to provide our health check script with a trusted CA via some means.

## Current result

Errors like this from the health check script and the Cloud Controller API getting deregistered.

```
{"timestamp":"2019-05-29T21:00:38.173610976Z","level":"info","source":"Route Registrar","message":"Route Registrar.Script exited with error","data":{"error":"exit status 60","script":"/var/vcap/jobs/cloud_controller_ng/bin/cloud_controller_ng_health_check","stderr":"  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n                                 Dload  Upload   Total   Spent    Left  Speed\n\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\r  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0\ncurl: (60) server certificate verification failed. CAfile: /etc/ssl/certs/ca-certificates.crt CRLfile: none\nMore details here: http://curl.haxx.se/docs/sslcerts.html\n\ncurl performs SSL certificate verification by default, using a \"bundle\"\n of Certificate Authority (CA) public keys (CA certs). If the default\n bundle file isn't adequate, you can specify an alternate file\n using the --cacert option.\nIf this HTTPS server uses a certificate signed by a CA represented in\n the bundle, the certificate verification probably failed due to a\n problem with the certificate (it might be expired, or the name might\n not match the domain name in the URL).\nIf you'd like to turn off curl's verification of the certificate, use\n the -k (or --insecure) option.\n","stdout":""}}
{"timestamp":"2019-05-29T21:00:38.173799415Z","level":"info","source":"Route Registrar","message":"Route Registrar.healthchecker returned unhealthy for route","data":{"route":{"Type":"","Name":"api","Port":9022,"TLSPort":9024,"Tags":{"component":"CloudController"},"URIs":["api.calico-crest.capi.land"],"RouterGroup":"","Host":"","ExternalPort":null,"RouteServiceUrl":"","RegistrationInterval":10000000000,"HealthCheck":{"Name":"api-health-check","ScriptPath":"/var/vcap/jobs/cloud_controller_ng/bin/cloud_controller_ng_health_check","Timeout":6000000000},"ServerCertDomainSAN":"api.calico-crest.capi.land"}}}
{"timestamp":"2019-05-29T21:00:38.173835217Z","level":"info","source":"Route Registrar","message":"Route Registrar.Unregistering route","data":{"route":{"Type":"","Name":"api","Port":9022,"TLSPort":9024,"Tags":{"component":"CloudController"},"URIs":["api.calico-crest.capi.land"],"RouterGroup":"","Host":"","ExternalPort":null,"RouteServiceUrl":"","RegistrationInterval":10000000000,"HealthCheck":{"Name":"api-health-check","ScriptPath":"/var/vcap/jobs/cloud_controller_ng/bin/cloud_controller_ng_health_check","Timeout":6000000000},"ServerCertDomainSAN":"api.calico-crest.capi.land"}}}
{"timestamp":"2019-05-29T21:00:38.173915992Z","level":"info","source":"Route Registrar","message":"Route Registrar.Unregistered routes successfully","data":{}}
```

## Possible Fix

Our current work around may be to just pass `-k` to the health check script's `curl` since validating TLS is tangential to what it's actually trying to do. Still, it would be nice for a more correct alternative here.

**Comment by ameowlia** [comment]
Hi @tcdowney, 

It looks like CAPI took on this work themselves. Thanks :D 

Can you confirm that you all have fixed this? 

Thanks, 
Amelia and Mikael @mike1808, CF Networking Program Members

**Comment by tcdowney** [comment]
@ameowlia @mike1808 

We didn't really do it ourselves, just kinda worked around it by configuring our health check script to curl using the `-k` flag.

https://github.com/cloudfoundry/capi-release/blob/e303bcc6bff610c72110d7c0ff352ce2d8706c47/jobs/cloud_controller_ng/templates/cloud_controller_api_health_check.erb#L19

It doesn't really matter a whole lot since we were just moving from hitting a regular `http` endpoint without TLS, but it would be nice if there was a way we could expose a trusted CA cert to our curl command.

**Comment by ameowlia** [comment]
@tcdowney our team is at low strength right now and I doubt we will be able to prioritize it soon. If this is important to you I would recommend submitting a PR. I think either solution you proposed would be fine, with slight preference for the BPM way.

Thanks 👋 
Amelia

**Comment by KauzClay** [comment]
Hi @tcdowney (@braa , @selzoc) , we did some digging and it looks like the entire job directory is mounted for the health check script in the route_registrar config. So if your health check really needs a cert, it can write it to disk and use it, because it will have access.

Thanks, 
@tcdowney and @KauzClay , CF Networking Program Members

**Comment by cwlbraa** [comment]
thanks for responding to yourself, @tcdowney ! exemplary dialogue! :tq:

**Comment by tcdowney** [comment]
Yes, cannot deviate from the CF Networking Program sign-off template™.

**Comment by mike1808** [comment]
Hey @tcdowney and @cwlbraa , sounds like this is resolved. We are gonna close it, but feel free to reopen.

Thanks,
@mike1808 and @KauzClay , CF Networking Program Members

**Comment by tcdowney** [comment]
Ok, sounds good. Thank you @mike1808 and @KauzClay. 🙂👍
